### PR TITLE
FINALLY reverse the k index

### LINKED ISSUE #90: Vertical index k increases as z coordinate decreases
For `RegularCartesianGrid`, `Oceananigans.jl` uses a convention in which indexing for the `z` coordinate in arrays is treated differently from `x` and `y`: the third index (call it `k`) *decreases* with increasing `z`, unlike the first and second indices which increase with increasing `x` and `y`.

This means `z` derivative operators look different from `x` and `y` derivative operators. For example, a `z` differential is defined as positive for the difference between `k` and `k+1`.

```Julia
@inline function δz_e2f(f, Nz, i, j, k)
    if k == Nz
        @inbounds return f[i, j, k]
    else
        @inbounds return f[i, j, k] - f[i, j, k+1]
    end
end
```

from [here](https://github.com/ali-ramadhan/Oceananigans.jl/blob/30150f965a1e0db95f239fa092ce33fe4ceedd29/src/operators/ops_regular_cartesian_grid_elementwise.jl#L38).

I recognize that some users, especially oceanographers, appreciate this feature in their output --- in particular because `k=1` corresponds to the 'surface', which has a special status in (some) oceanographic applications. 

However, I believe this 'reverse indexing' convention for the third direction is potentially confusing to developers and users with backgrounds in physics or engineering and may lead to bugs, slow down development, and stymie super-users who need to write code that indexes into arrays.

I propose that we consider using the convention where indices increase with increasing coordinates in all directions (for all coordinate systems), so that for a Cartesian coordinate system `z` and `k` increase upwards, like 'x', 'y', 'i' and 'j'. I believe this behavior is probably expected for the majority of developers of fluid codes in general, and also for a large fraction of users.

To satisfy users who prefer the oceanographic indexing convention in their output, we can provide users with the option (and make it default if we wish) to flip output and grids on their head before saving, so that `k=1` corresponds to the surface *in output*.

@ali-ramadhan @christophernhill @jm-c @SandreOuza

### LINKED ISSUE #468: Rename Pearson vortex test to Taylor-Green vortex
Overleaf documentation reflects this but not the code itself.

This is the analytic solution we're testing against: https://en.wikipedia.org/wiki/Taylor%E2%80%93Green_vortex#Taylor%E2%80%93Green_vortex_solution

### LINKED ISSUE #480: Round-off error in `RegularCartesianGrid` produces ranges of wrong length
Noticed this because `test_diffusion_cosine` was failing in PR #462.

We should have that `length(zC) == Nz` but
```julia
julia> Nz = 16; Lz = π/2; Δz = Lz / Nz; zC = -Lz+Δz/2:Δz:-Δz/2; length(zC)
16

julia> Nz = 32; Lz = π/2; Δz = Lz / Nz; zC = -Lz+Δz/2:Δz:-Δz/2; length(zC)
32

julia> Nz = 64; Lz = π/2; Δz = Lz / Nz; zC = -Lz+Δz/2:Δz:-Δz/2; length(zC)
63

julia> Nz = 128; Lz = π/2; Δz = Lz / Nz; zC = -Lz+Δz/2:Δz:-Δz/2; length(zC)
127

julia> Nz = 256; Lz = π/2; Δz = Lz / Nz; zC = -Lz+Δz/2:Δz:-Δz/2; length(zC)
255
```

Looks like a scary effect of round-off error to me.

I'll make the switch to using the `range` function with a fixed `length` and add a test to check for this.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR finally reverses the `k` index although it's still a work in progress as a few tests fail. In particular, it seems that incompressibility is not satisfied even though I managed to get the "recomputing w from continuity" test to pass.

Need to think about this some more.

Resolves #90
Resolves #468 
Resolves #480

### CODE REVIEW

**Comment by glwagner** [comment]
Are we integrating bottom-up or top-down now?

Ultimately it might be nice to eliminate that step and explicitly time-step the vertical velocity, since thats valid for general boundary conditions...

File: src/time_steppers.jl [hunk 1]
```diff
@@ -410,9 +410,9 @@ Compute the vertical velocity w by integrating the continuity equation downwards
 function compute_w_from_continuity!(grid::AbstractGrid, U)
     @loop for j in (1:grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
         @loop for i in (1:grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
-            @inbounds U.w[i, j, 1] = 0
-            @unroll for k in 2:grid.Nz
-                @inbounds U.w[i, j, k] = U.w[i, j, k-1] + grid.Δz * ∇h_u(i, j, k-1, grid, U.u, U.v)
+            @inbounds U.w[i, j, grid.Nz] = 0
```
Comment (glwagner) [inline src/time_steppers.jl]: This is incorrect --- remember that `Nz` is inside the domain. 

I think it'd be clearer to integrate from the bottom upwards; eg start at `k=1` and integrate up to `Nz`.

Otherwise, you should set the value of `w` at `Nz+1` (the top surface), and integrate downwards.

Comment (glwagner) [reply src/time_steppers.jl]: To integrate upwards, write

```julia
U.w[i, j, 1] = 0
@unroll for k in 2:grid.Nz
    @inbounds U.w[i, j, k] = U.w[i, j, k-1] - grid.Δz * ∇h_u(i, j, k, grid, U.u, U.v)
end
```

Comment (glwagner) [reply src/time_steppers.jl]: In summary, I think this function will work either way as long as you ensure that

1. You integrate over *all* interior grid points --- either `2:grid.Nz`, or `grid.Nz:-1:2` --- and

2. You refrain from setting the value of `w` at either top or bottom, since ensuring the correct value of `w` along the boundary is the job of `fill_halo_regions!` 

(Point 2 is not strictly necessary at this point since we only support `NoPenetration` boundary conditions, but will mean that adding new `Periodic` / Flow` / `Penetration` z-boundary conditions for vertical velocity will work out of the box without having to change this function.)

**Comment by ali-ramadhan** [comment]
For w is
1. `k = Nz` now the surface implying that `k=0` is the bottom and a halo cell?
2. or is `k = 1` the bottom implying that `k = Nz+1` is the surface and a halo cell?

Seems that we're actually implementing the second.

Might be nice to have direct access to the surface via `k = Nz` but this would break the pattern that `i = 1` for u and `j = 1` for v coincide with the boundary whereas `i = Nx` and `j = Ny` do not.

Will ask @jm-c about this.

**Comment by glwagner** [comment]
The convention we use for the other variables stored on faces is that our "interior" data spans from `i=1` to `i=N`. In addition, we use a convention that face indices are "left" of cell indices. Thus `iFace=1` corresponds to the "left" boundary of the domain. `iCell=N` is the rightmost cell in the domain; and `iFace=N+1` corresponds to the rightmost boundary.

Why do you want "direct" access to the surface at `k=Nz`? I don't understand.

File: src/time_steppers.jl (hunk 1 above)
Comment (glwagner) [reply src/time_steppers.jl]: I think we should be able to explain why integrating from the bottom up does not work.

**Comment by ali-ramadhan** [comment]
> Why do you want "direct" access to the surface at `k=Nz`? Can you explain why you want this?

We've mostly been concerned with surface ocean physics so naively it would seem nice to have the surface not be a halo cell.

Might have been an issue in the past when we didn't have vertical halos because then you don't even have a cell for the surface.

Out of curiousity then, would an implementation of surface wave effects (#443) have to touch the halos?

**Comment by glwagner** [comment]
> We've mostly been concerned with surface ocean physics so naively it would seem nice to have the surface not be a halo cell.

This seems like a question of indexing rather than halos, but maybe thats not true given that we use the halos more as ghost points now, rather than overlap regions for MPI stuff, etc.

File: src/time_steppers.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/time_steppers.jl]: I agree. Perhaps I opened this PR too early as I'm still debugging. Will revisit once I've worked on other parts of this PR.

Comment (ali-ramadhan) [reply src/time_steppers.jl]: @jm-c suggested that integrating from the bottom upwards is preferable as it will also work in the case that we decide to implement a free surface.

**Comment by ali-ramadhan** [comment]
Fixed a few more things and I've updated `recompute_w_from_continuity!` so that the integral is computed starting from the bottom upwards.

Regression tests fail so I'll see what's wrong there tomorrow.

I also used `reverse(; dims=3)` to flip the arrays in the regression tests.

We should probably go in this order:
1. Merge LES regression tests (PR #478)
2. Merge arbitrary tracers (PR #452)
3. Make sure this PR passes the LES regression tests and works with arbitrary tracers.

File: src/grids.jl [hunk 2]
```diff
@@ -103,13 +103,13 @@ function RegularCartesianGrid(T, N, L)
 
     V = Δx*Δy*Δz
 
-    xC = Δx/2:Δx:Lx
-    yC = Δy/2:Δy:Ly
-    zC = -Lz+Δz/2:Δz:-Δz/2
+    xC = Nx == 1 ? (Δx/2:Δx:Lx) : range(Δx/2, Lx-Δx/2; length=Nx)
```
Comment (glwagner) [inline src/grids.jl]: Why the if statement? I get

```julia
julia> a = range(1.0, 1.0, length=1)
1.0:0.0:1.0
```

which seems right.

Comment (ali-ramadhan) [reply src/grids.jl]: Hmmm, when I first made the change I hit an error like
```julia
julia> range(0.5, 1; length=1)
ERROR: ArgumentError: range(0.5, stop=1.0, length=1): endpoints differ
```
so I decided to reproduce the previous behavior when `N == 1` but maybe I fixed that...

Doing this manually seems to work, so I'll try switching back to `range`.

**Comment by ali-ramadhan** [comment]
Regression tests all pass on my laptop!

Next I'll merge in master to make sure `reverse-k-index` works with arbitrary tracers and can pass the LES regression tests.

Then we can regenerate the regression data with a reversed `k` index.

**Comment by ali-ramadhan** [comment]
After reversing the k index u, v, w, and S are all correct if you take one time step for the LES regression tests (non-LES regression tests all pass). But T values at the top and bottom grid cells are wrong.

Still thinking of what could be effecting T but not S (which also runs with AMD and flux boundary conditions)...

T is the only field with a gradient boundary condition, but enforcing gradient boundary conditions shouldn't have changed and looks fine to me...

I think this is the last bug, if fixed this should become ready to merge.

File: src/halo_regions.jl [hunk 3]
```diff
@@ -24,12 +24,12 @@ _fill_south_halo!(c, ::PBC, H, N) = @views @. c.parent[:, 1:H, :] = c.parent[:,
  _fill_north_halo!(c, ::PBC, H, N) = @views @. c.parent[:, N+H+1:N+2H, :] = c.parent[:, 1+H:2H, :]
 _fill_bottom_halo!(c, ::PBC, H, N) = @views @. c.parent[:, :, N+H+1:N+2H] = c.parent[:, :, 1+H:2H]
```
Comment (glwagner) [inline src/halo_regions.jl]: `_fill_top_halo` and `_fill_top_halo` for flux and periodic boundary conditions need to be swapped: the "bottom" grid points range on the parent array from `1:H`, for example.

**Comment by ali-ramadhan** [comment]
Thanks @glwagner! All four regression tests pass now!

I'll regenerate the regression data in a subsequent PR which also remove all the `reverse` statements.

**Comment by ali-ramadhan** [comment]
Differences are close to machine epsilon so pretty sure this branch produces the same numbers as before, up to truncation error (and we do integrate from the bottom up to enforce continuity now).

```
Running regression tests...
  Testing thermal bubble regression [CPU]
[ Info: Δu: min=-3.01224e-15, max=3.01289e-15, mean=-3.94216e-22, absmean=5.13638e-16, std=7.36011e-16
[ Info: Δv: min=-3.01208e-15, max=3.00899e-15, mean=-1.19521e-21, absmean=5.13326e-16, std=7.35825e-16
[ Info: Δw: min=-7.64756e-15, max=5.88619e-15, mean=1.35968e-21, absmean=1.03765e-15, std=1.48909e-15
[ Info: ΔT: min=-1.77636e-15, max=1.77636e-15, mean=7.80626e-18, absmean=3.55618e-17, std=2.51247e-16
[ Info: ΔS: min=0, max=0, mean=0, absmean=0, std=0
  Testing Rayleigh–Bénard tracer regression [CPU]
[ Info: Δu: min=-3.65818e-14, max=3.21965e-14, mean=-1.92406e-18, absmean=3.3529e-15, std=4.83978e-15
[ Info: Δv: min=-2.85535e-14, max=3.72202e-14, mean=1.62974e-19, absmean=3.29709e-15, std=4.79574e-15
[ Info: Δw: min=-4.56857e-14, max=4.32987e-14, mean=2.54957e-18, absmean=2.97761e-15, std=4.68161e-15
[ Info: ΔT: min=-5.19029e-14, max=3.71925e-14, mean=8.223e-18, absmean=2.55428e-15, std=4.14735e-15
[ Info: ΔS: min=-2.37588e-14, max=1.60982e-14, mean=1.15112e-18, absmean=1.13258e-15, std=1.79417e-15
  Testing oceanic large eddy simulation regression [VerstappenAnisotropicMinimumDissipation, CPU]
[ Info: Δu: min=-8.1532e-16, max=7.47666e-16, mean=3.10996e-19, absmean=1.34093e-16, std=1.74514e-16
[ Info: Δv: min=-7.94503e-16, max=7.32053e-16, mean=1.62668e-19, absmean=1.30967e-16, std=1.70133e-16
[ Info: Δw: min=-2.21871e-15, max=2.4928e-15, mean=-4.54454e-18, absmean=2.67538e-16, std=4.09625e-16
[ Info: ΔT: min=-1.06581e-14, max=1.06581e-14, mean=-1.76661e-17, absmean=1.34445e-15, std=2.4561e-15
[ Info: ΔS: min=-2.13163e-14, max=2.13163e-14, mean=-9.38131e-17, absmean=2.61336e-15, std=4.82786e-15
  Testing oceanic large eddy simulation regression [SmagorinskyLilly, CPU]
[ Info: Δu: min=-7.65013e-16, max=8.1532e-16, mean=-4.16382e-20, absmean=1.23114e-16, std=1.62955e-16
[ Info: Δv: min=-7.56339e-16, max=8.43076e-16, mean=-1.08495e-19, absmean=1.21821e-16, std=1.61572e-16
[ Info: Δw: min=-2.19616e-15, max=1.9585e-15, mean=6.95974e-18, absmean=2.40883e-16, std=3.76408e-16
[ Info: ΔT: min=-1.06581e-14, max=1.06581e-14, mean=-9.13764e-18, absmean=1.30546e-15, std=2.43045e-15
[ Info: ΔS: min=-2.13163e-14, max=2.13163e-14, mean=-1.34019e-17, absmean=2.589e-15, std=4.83055e-15
```

File: src/Oceananigans.jl [hunk 4]
```diff
@@ -49,7 +49,7 @@ export
 
     # Model output writers
     Checkpointer, restore_from_checkpoint, read_output,
-    JLD2OutputWriter, FieldOutput, FieldOutputs,
+    JLD2OutputWriter, NetCDFOutputWriter, FieldOutput, FieldOutputs,
```
Comment (glwagner) [inline src/Oceananigans.jl]: ?

File: src/boundary_conditions.jl [hunk 5]
```diff
@@ -214,7 +211,7 @@ function ChannelBCs(;  north = BoundaryCondition(Flux, nothing),
 
     x = PeriodicBCs()
     y = CoordinateBoundaryConditions(south, north)
-    z = CoordinateBoundaryConditions(top, bottom)
+    z = CoordinateBoundaryConditions(bottom, top)
```
Comment (glwagner) [inline src/boundary_conditions.jl]: Very satisfying.

File: test/test_turbulence_closures.jl [hunk 6]
```diff
@@ -136,9 +135,8 @@ function test_function_differentiation(T=Float64)
     ∂y_ϕ_f = ϕ²[2, 2, 2] - ϕ²[2, 1, 2]
     ∂y_ϕ_c = ϕ²[2, 3, 2] - ϕ²[2, 2, 2]
 
-    # Note reverse indexing here!
```
Comment (glwagner) [inline test/test_turbulence_closures.jl]: Nice.

File: src/Oceananigans.jl (hunk 4 above)
Comment (ali-ramadhan) [reply src/Oceananigans.jl]: Yeah this was an issue after I merged master in, I guess we forgot to export it in PR #433.

Bad practice on my part to include changes not relevant to reversing the k index.
